cpp/src/fil/fil.cu

```diff
@@ -436,6 +456,9 @@ struct sparse_forest : forest {
     nodes_.resize(num_nodes_, h.get_stream());
     CUDA_CHECK(cudaMemcpyAsync(
       nodes_.data(), nodes, sizeof(node_t) * num_nodes_, cudaMemcpyHostToDevice, h.get_stream()));
+
+    dispatch_on_fil_template_params(enable_smem_carveout<sparse_storage<node_t>>(max_shm_),
+                                    (predict_params)class_ssp_);
```

Comment:
Why is a type cast needed here?

Reply:
because the type is used to resolve a template instantiation, and for template resolutions, such conversions are not considered.